[JENKINS-59480] support for SUSE

mawinter69 opened the pull request:
## [JENKINS-59480](https://issues.jenkins-ci.org/browse/JENKINS-59480) - support for SUSE

Older SUSE release (until 12) have a file /etc/SuSE-release.  /etc/os-release only exists from SLES 12 onwards. On SLES11 and earlier the lsb_release only provides the major version but not the patch level.  This change adds support for reading the SuSE-release file and falling back to it on SLES11 and older to get better vesrion information. It will also map "sles" from /etc/os-release to "SUSE LINUX"

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/jenkinsci/platformlabeler-plugin/blob/master/CONTRIBUTING.md) doc
- [x] I have referenced the Jira issue related to my changes in one or more commit messages
- [x] I have added tests that verify my changes
- [x] Unit tests pass locally with my changes
- [x] I have added documentation as necessary
- [x] No spotbugs warnings were introduced with my changes
- [x] I have interactively tested my changes

## Types of changes

- [x] New feature (non-breaking change which adds functionality)

## Further comments

Up to SLES12 SP1, lsb_release returns "SUSE LINUX" from 12.2 onwards only "SUSE" is returned. Should this be unified to always return "SUSE" or "SUSE LINUX"?

This change is breaking in the sense that before when only reading /etc/os-release, the platform name would have been "sles", now it will be "SUSE".

In case there will be more platforms that need special handling it might make sense to have an extension point for this.

MarkEWaite commented on README.md:
```diff
@@ -22,6 +22,10 @@ Labels commonly include operating system name, version, and architecture.
 | Red Hat Enterprise Linux 8 | `RedHatEnterprise` | `8.0`          | `amd64`  |
 | Scientific 6.10            | `Scientific`       | `6.10`         | `amd64`  |
 | Scientific 7.7             | `Scientific`       | `7.7`          | `amd64`  |
+| SLES 11.3                  | `SUSE LINUX`       | `11.3`          | `amd64`  |
```
Labels that contain embedded spaces are much more difficult to manage than labels without embedded spaces.  Since the lsb_release `Distributor ID` field is defined with rules that preclude a space in the ID, I'd like this to be `SUSE` rather than `SUSE LINUX`.
```suggestion
| SLES 11.3                  | `SUSE`             | `11.3`          | `amd64`  |
```

MarkEWaite commented:
Thanks so much for this contribution!  I would never have been able to cover all those versions of SuSE Linux.

MarkEWaite commented:
> Up to SLES12 SP1, lsb_release returns "SUSE LINUX" from 12.2 onwards only "SUSE" is returned. Should this be unified to always return "SUSE" or "SUSE LINUX"?

SUSE rather than SUSE LINUX because label management is easier when the labels do not include spaces and because I prefer the newer form of label rather than the older form of label.

> This change is breaking in the sense that before when only reading /etc/os-release, the platform name would have been "sles", now it will be "SUSE".

Since processing of `/etc/os-release` is a very recent addition ([June 15, 2019 release 3.1](https://github.com/jenkinsci/platformlabeler-plugin/releases/tag/platformlabeler-3.1)), I accept that incompatible behavior change.

mawinter69 commented:
Changed it to always return SUSE and modified the pom to report the compatibility for the update site. The exact version needs to be changed to whatever is the used version during the release where this change is included.

MarkEWaite commented:
> Changed it to always return SUSE and modified the pom to report the compatibility for the update site. The exact version needs to be changed to whatever is the used version during the release where this change is included.

Thanks.  I think we should declare this to be version 5.0.  Major version increment because this is a significant addition (SUSE Linux support) and because there is an intentional compatibility break.  Does that work OK for you?

mawinter69 commented:
sure 5.0 is fine for me

MarkEWaite commented:
@mawinter69 was there a reason you chose to mark the incompatibility using the older technique rather than the single line entry in the [properties section](https://wiki.jenkins.io/display/JENKINS/Marking+a+new+plugin+version+as+incompatible+with+older+versions)?
